core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -108,31 +111,39 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
   /**
    * Sets topic partition states for the given partitions.
    * @param leaderIsrAndControllerEpochs the partition states of each partition whose state we wish to set.
+   * @param controllerEpochVersion expected controller epoch zkVersion to check.
    * @return sequence of SetDataResponse whose contexts are the partitions they are associated with.
    */
-  def setTopicPartitionStatesRaw(leaderIsrAndControllerEpochs: Map[TopicPartition, LeaderIsrAndControllerEpoch]): Seq[SetDataResponse] = {
-    val setDataRequests = leaderIsrAndControllerEpochs.map { case (partition, leaderIsrAndControllerEpoch) =>
+  def setTopicPartitionStatesRaw(leaderIsrAndControllerEpochs: Map[TopicPartition, LeaderIsrAndControllerEpoch], controllerEpochVersion: Int): Seq[MultiOpResponse] = {
+    val multiOpRequests = leaderIsrAndControllerEpochs.map { case (partition, leaderIsrAndControllerEpoch) =>
```

Comment:
Here `SetDataRequest/SetDataResponse` is replaced with `MultiOpRequest/MultiOpResponse`. This may cause problem for existing code (e.g. `ZookeeperClient.send()`) whose logic relies on the type of the request.

Instead of adding a new subclass of AsyncRequest, would it be better to modify the existing request (maybe the AsyncRequest) to include the expected controller epoch version so that, when the expected controller epoch exists, zookeeperClient.send() will take care of the version check?

Reply:
That is a good point. The reason why I add the MultiOp subclass is that we can extend `ZookeeperClient` to handle arbitrary multi() operations, not specifically for checking controller epoch version and updating zk states. Also, I think `ZookeeperClient` should only be aware of zookeeper related context not the kafka related context (e.g. controller epoch version) for cleanness. Instead of modifying AsyncRequest, do you think it is better to add some helper functions in `KafkaZkClient` to wrap around the check and set/update/create logic?

Also, I am a little bit confused on what are the problems caused by `ZookeeperClient.send()` if we use `zookeeper.multi()` for `MultiOpRequest`. Can you give me more contexts on that?

Reply:
I think it is reasonable not to have kafka specific thing in ZookeeperClient. On the other hand we don't have to -- we can provide anther zk path and expected version as parameters to these APIs, the API should return proper error without executing the original request if the version of the path is different from the expected version. This solution is probably not Kafka specific.

Not sure that we don't need arbitrary multi() operations as of now. Currently we only need to check the zk version of another path when controller changes zookeeper state.

After checking the code, it seems that there is no current problem caused by using MultiOpRequest. But some information (e.g. `CreateResponse.name` and `SetDataResponse.stat`) is discarded in the response which may potentially be problem in the future.  In general it seems more flexible to be able to use different case class for different requests so that we can have different parameters (as is the case now) and apply different processing logic to different case class. Just my opinion. If you like the current solution, maybe you can keep it and other committers can comment on this.

Reply:
I see what you mean. That makes sense to me. I will update the PR to differentiate Set/Update/Create with check parameters provided in `ZookeeperClient` to expose different information. Thanks for the explanation.